libcrux-psq/src/handshake/ciphersuite.rs

```diff
@@ -0,0 +1,747 @@
+use libcrux_kem::{MlKem768Ciphertext, MlKem768PrivateKey, MlKem768PublicKey};
+use libcrux_ml_kem::MlKemSharedSecret;
+use rand::CryptoRng;
+use tls_codec::{
+    Deserialize, Serialize, SerializeBytes, Size, TlsDeserialize, TlsSerialize, TlsSerializeBytes,
+    TlsSize,
+};
+
+#[cfg(feature = "classic-mceliece")]
+use crate::classic_mceliece::{Ciphertext, PublicKey, SecretKey, SharedSecret};
+use crate::handshake::{
+    dhkem::{DHKeyPair, DHPrivateKey, DHPublicKey},
+    HandshakeError,
+};
+
+#[derive(Clone, Copy, Debug)]
+pub enum CiphersuiteName {
+    X25519ChachaPolyHkdfSha256,
+    X25519Mlkem768ChachaPolyHkdfSha256,
+    X25519ClassicMcElieceChachaPolyHkdfSha256,
+}
+
+pub struct CiphersuiteBuilder<'a> {
+    longterm_ecdh_keys: Option<&'a DHKeyPair>,
+    longterm_mlkem_encapsulation_key: Option<&'a MlKem768PublicKey>,
+    longterm_mlkem_decapsulation_key: Option<&'a MlKem768PrivateKey>,
+    peer_longterm_ecdh_pk: Option<&'a DHPublicKey>,
+    peer_longterm_mlkem_pk: Option<&'a MlKem768PublicKey>,
+    #[cfg(feature = "classic-mceliece")]
+    longterm_cmc_encapsulation_key: Option<&'a PublicKey>,
+    #[cfg(feature = "classic-mceliece")]
+    longterm_cmc_decapsulation_key: Option<&'a SecretKey>,
+    #[cfg(feature = "classic-mceliece")]
+    peer_longterm_cmc_pk: Option<&'a PublicKey>,
+}
+
+#[allow(non_camel_case_types)]
+pub enum InitiatorCiphersuite<'a> {
+    X25519_ChaChaPoly_HkdfSha256(InitiatorX25519ChachaPolyHkdfSha256<'a>),
+    X25519_MlKem768_ChaChaPoly_HkdfSha256(InitiatorX25519Mlkem768ChachaPolyHkdfSha256<'a>),
+    #[cfg(feature = "classic-mceliece")]
+    X25519_ClassicMcEliece_ChaChaPoly_HkdfSha256(
+        InitiatorX25519ClassicMcElieceChachaPolyHkdfSha256<'a>,
+    ),
+    #[cfg(not(feature = "classic-mceliece"))]
+    X25519_ClassicMcEliece_ChaChaPoly_HkdfSha256(DummyCiphersuite),
+}
+
+#[allow(non_camel_case_types)]
+pub enum ResponderCiphersuite<'a> {
+    X25519_ChaChaPoly_HkdfSha256(ResponderX25519ChaChaPolyHkdfSha256<'a>),
+    X25519_MlKem768_ChaChaPoly_HkdfSha256(ResponderX25519MlKem768ChaChaPolyHkdfSha256<'a>),
+    #[cfg(feature = "classic-mceliece")]
```

Comment:
Not the nicest, but I'm not sure what kind of trouble this may still be able to do. So I think this is fine. When everything is boxed it may also be equivalent in memory.